### PR TITLE
Add component type to dependencies to self

### PR DESCRIPTION
**Pull request by codecholeric**
This is the reverse side of #428, i.e. it adds the component type of arrays to `JavaClass.dependenciesToSelf`. In other words it for the following example

```
class Target {}

class DependsOnComponentType {
    Target[] fieldDependency;
}
```

we will now see the dependency `DependsOnComponentType -> Target` for `target.getDependenciesToSelf()`.

On the way I figured that the dependency resolution process, and in particular the reverse dependency resolution process, is meanwhile pretty convoluted. I have thus simplified this and at least on my machine increased the speed for the reverse dependency lookup by a factor 7 for `ClassFileImporterSlowTest.importsClasspath()`.
In the end I figured that the process of adding new types of dependencies is pretty tedious, because we first add the "dependencies from self", then in a completely parallel process we add the "dependencies to self". With the refactoring now we will simply reuse the "dependencies from self" to create the "dependencies to self", thus making it a lot easier in the future to add both directions in one shot.

Issue: #257

### CODE REVIEW

File: archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClass.java [hunk 1]
```diff
@@ -1048,55 +1072,71 @@ public JavaConstructor getConstructor(String... parameters) {
...
-        return javaClassDependencies.getAnnotationsWithTypeOfClass();
+        return reverseDependencies.getAnnotationsWithTypeOf(this);
+    }
+
+    /**
+     * @return All imported {@link JavaAnnotation JavaAnnotations} that have a parameter with type of this class.
+     */
+    @PublicAPI(usage = ACCESS)
+    public Set<JavaAnnotation<?>> getAnnotationsWithParameterTypeOfSelf() {
+        return reverseDependencies.getAnnotationsWithParameterTypeOf(this);
```
Comment (hankem) [inline archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClass.java:1131]: According to the test coverage, this can – as well as `getInstanceofChecksWithTypeOfSelf` – _just be simplified_ to:
```
return Collections.emptySet();
```
Maybe we should have some tests for those new methods...? 😜

**Review by hankem — COMMENTED** [review]
I cannot claim (at all) that I had understood how all the `AccessContext`, `TopProcess`, `CompletionProcess`, `MemberDependenciesByTarget` worked before... 🙈
... but I definitely agree that the new concept of `ReverseDependencies` is much simpler! 👍
It would take me much more time to understand everything in detail, but IMO the Christmas release shouldn't be blocked by that.

Regarding the speedup: You were only looking at _the reverse dependency lookup_, right? (What did you measure exactly?)
For the entire import (or even execution) of the `ClassFileImporterSlowTest`, I didn't see a significant difference, nor for the class file import in some larger project.

**Comment by codecholeric** [comment]
Thanks a lot for looking into it!! Yes, this is probably one of the most complicated parts :wink: If you understand the code after better than before, I would say mission accomplished :joy: I mean, there is IMHO a really good test coverage, so I'm not too scared that I have destroyed anything fundamental with this refactoring. But  there can always be some subtle corner cases I've overlooked.

Regarding performance, yes, the dependency and reverse dependency creation is a lazy process. So what I did to measure the difference is to add the following to `ClassFileImporterSlowTest.importsClasspath()`

```
for (JavaClass class : classes) {
    class.getDirectDependenciesToSelf();
}
```

If you don't do anything like that, then all the changed code is not executed anyway :wink:

File: archunit/src/main/java/com/tngtech/archunit/core/importer/AccessRecord.java [hunk 2]
```diff
@@ -273,17 +282,59 @@ private static JavaCodeUnit getCaller(CodeUnit caller, ImportedClasses classes)
                     " that matches supposed caller " + caller);
         }
 
-        private static <T extends HasName & HasDescriptor & HasOwner<JavaClass>> Set<T>
-        tryFindMatchingTargets(Set<T> possibleTargets, TargetInfo targetInfo) {
-            ImmutableSet.Builder<T> result = ImmutableSet.builder();
-            for (T possibleTarget : possibleTargets) {
-                if (targetInfo.matches(possibleTarget)) {
+        private static <MEMBER extends JavaMember, TARGET extends TargetInfo> Set<MEMBER>
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/importer/AccessRecord.java:285]: :see_no_evil:

File: archunit/src/main/java/com/tngtech/archunit/core/domain/ReverseDependencies.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        this.methodParameterTypeDependencies = methodParameterTypeDependencies;
+        this.methodReturnTypeDependencies = methodReturnTypeDependencies;
+        this.methodsThrowsDeclarationDependencies = methodsThrowsDeclarationDependencies;
+        this.constructorParameterTypeDependencies = constructorParameterTypeDependencies;
+        this.constructorThrowsDeclarationDependencies = constructorThrowsDeclarationDependencies;
+        this.annotationTypeDependencies = annotationTypeDependencies;
+        this.annotationParameterTypeDependencies = annotationParameterTypeDependencies;
+        this.instanceofCheckDependencies = instanceofCheckDependencies;
+        this.directDependenciesToClass = createDirectDependenciesToClassSupplier(allDependencies);
+    }
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/domain/ReverseDependencies.java:77]: Agreed, this also matches the way the builder pattern is usually implemented within ArchUnit :smiley:

File: archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClass.java (hunk 1 above)
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClass.java:1131]: Ah yes, now I remember that I wanted to add some additional test :see_no_evil: In particular, because for the final state this method is not used anymore to create the reverse dependencies :wink:

File: archunit/src/main/java/com/tngtech/archunit/core/importer/AccessRecord.java [hunk 4]
```diff
@@ -273,17 +282,59 @@ private static JavaCodeUnit getCaller(CodeUnit caller, ImportedClasses classes)
...
+            return target.owner.getFullyQualifiedClassName().equals(member.getOwner().getName()) ||
+                    containsExactlyOneMatch(new ClassHierarchyPath(target.owner, member.getOwner()), target, signaturePredicate);
+        }
+
+        private static <TARGET extends TargetInfo> boolean containsExactlyOneMatch(Iterable<JavaClass> classes, TARGET target, SignaturePredicate<TARGET> signaturePredicate) {
+            Set<JavaClass> matching = new HashSet<>();
+            for (JavaClass javaClass : classes) {
+                if (signaturePredicate.exists(javaClass, target)) {
+                    matching.add(javaClass);
+                }
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/importer/AccessRecord.java:309]: Yeah, this only occurs, if a method has been overridden several times in the hierarchy. E.g. if you have

```
class Base {
  void foo() {
    // some impl
  }
}

class Child extends Base {
  @Override
  void foo() {
    // also some impl
  }
}

class GrandChild extends Child {
  // not overridden here, otherwise it would be directly found
}
```

And then by accident you pick `Base.foo()` as the first candidate to find the target of a call to `GrandChild.foo()`. I'll still look into it :+1: (unfortunately the whole "start with any method in the hierarchy is pretty sub-optimal" :disappointed:)

[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/importer/AccessRecord.java:309]: Hmm, I think I'll leave it out for now. I could not find any noticeable difference in performance. And the problem is, that the benefit is not completely clear for me. We are exchanging a couple of unnecessary loops in a rare case against an operation we have to execute in every single loop. Take the extreme case that we for example have no single method that has been overridden multiple times in the whole set of imported classes, then we would actually even perform worse than if we don't do the check.

[the remark this answers is not in the document]
Comment (hankem) [reply archunit/src/main/java/com/tngtech/archunit/core/importer/AccessRecord.java:309]: I absolutely agree.
When I initially thought about it, I hadn't realized what `containsExactlyOneMatch` is actually used for...

File: archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClass.java (hunk 1 above)
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClass.java:1141]: Okay, I added tests for the two new public API methods :smiley:
